File: MK-52_ESP32/src/Receiver_PROG_A.cpp
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

#include "Receivers.hpp"
#include "../functions/functions.h"

#define __DEBUG

using namespace MK52_Interpreter;

void Receiver_PROG_A::init( void *components[]) {
    #ifdef __DEBUG
    Serial.println( "Init PROG_A");
    #endif
    _ar = (Receiver_Address *)components[COMPONENT_RECEIVER_ADDRESS];
    Receiver::init(components);
}

void Receiver_PROG_A::activate( uint8_t scancode, int8_t parent){
    #ifdef __DEBUG
    Serial.println( "Activating receiver PROG_A");
    #endif
    Receiver::activate( scancode, parent);
    _lcd->updateStatusFMODE( " A ");
    _mode = 1;
    if(!scancode) return;
    tick(scancode);
}

int Receiver_PROG_A::tick( uint8_t scancode){    int return_value = COMPONENT_RECEIVER_PROG_N;
    if( _ar->isActive()){        
        if( _ar->tick( scancode) == NO_CHANGE){
            sprintf_P( _rpnf->getOutputBuffer(), PSTR("%04u> "), _rpnf->progMem->getCounter());
            _rpnf->appendOutputBuffer( _rpnf->progMem->getCurrentLine());
            _rpnf->appendOutputBuffer( _ar->toTrimmedString());
            _lcd->updateTerminalLine( 10, _rpnf->getOutputBuffer());
            return NO_CHANGE;
        }
        _rpnf->setOutputBuffer( _rpnf->progMem->getCurrentLine());
        _rpnf->appendOutputBuffer( _ar->toString());
        _rpnf->progMem->replaceLine( _rpnf->getOutputBuffer());
        _rpnf->progMem->incrementCounter();
        _lcd->updateStatusPC( _rpnf->progMem->getCounter());
        return return_value;
    }
    switch( scancode){
        // Column 0
        case 1:
            _mode = 0;
            return COMPONENT_RECEIVER_PROG_F;
        case 2:
            _mode = 0;
            return COMPONENT_RECEIVER_PROG_K;
        case 4:
            _rpnf->execute(FUNC_TOGGLE_EMOD);
            _lcd->updateStatusDMODE(_rpnf->progMem->getEModeName());
            return NO_CHANGE;

        // Column 1
        case 5:
            _rpnf->appendProgramLine_P(FUNC_IFNOTLTY);
            _ar->activate( 0, COMPONENT_RECEIVER_PROG_N);
            return NO_CHANGE;
        case 6:
            _rpnf->appendProgramLine_P(FUNC_IFNOTEQY);
            _ar->activate( 0, COMPONENT_RECEIVER_PROG_N);
            return NO_CHANGE;
        case 7:
            _rpnf->appendProgramLine_P(FUNC_IFNOTGEY);
            _ar->activate( 0, COMPONENT_RECEIVER_PROG_N);
            return NO_CHANGE;
        case 8:
            _rpnf->appendProgramLine_P(FUNC_IFNOTNEY);
            _ar->activate( 0, COMPONENT_RECEIVER_PROG_N);
            return NO_CHANGE;

        // Column 2
        case 9:
            _rpnf->appendProgramLine_P(FUNC_A_X2M);
            _ar->activate( 0, COMPONENT_RECEIVER_PROG_N);
            return NO_CHANGE;
        case 10:
            _rpnf->appendProgramLine_P(FUNC_A_M2X);
            _ar->activate( 0, COMPONENT_RECEIVER_PROG_N);
            return NO_CHANGE;

        // Column 3 does nothing
        // Column 4 does nothing

        // Column 5
        case 21:
            // TODO
            return NO_CHANGE;
        case 22:
            _rpnf->appendProgramLine_P(FUNC_D2RAD);
            break;
        case 23:
            _rpnf->appendProgramLine_P(FUNC_MM2IN);
            break;
        case 24:
            #ifdef __DEBUG
            Serial.println("Going to FILE Display");
            #endif
            _mode = 0;
            return COMPONENT_DISPLAY_FILE;

        // Column 6
        case 25:
            // TODO
            return NO_CHANGE;
        case 26:
            _rpnf->appendProgramLine_P(FUNC_RAD2D);
            break;
        case 27:
            _rpnf->appendProgramLine_P(FUNC_IN2MM);
            break;
        case 28:
            #ifdef __DEBUG
            Serial.println("Going to DATA Display");
            #endif
            _mode = 0;
            return COMPONENT_DISPLAY_DATA;

        // Column 7
        case 29:
            // TODO
        case 30:
            // TODO
            return NO_CHANGE;
        case 31:
            _rpnf->appendProgramLine_P(FUNC_SEED);
            break;
        case 32:
            _mode = 0;
            return SHUTDOWN_REQUESTED;

        default: // all other buttons do nothing - keep A-mode
           return NO_CHANGE;
    }
    _mode = 0;
    _rpnf->progMem->incrementCounter();
    _lcd->updateStatusPC( _rpnf->progMem->getCounter());
    return return_value;
}

File: MK-52_Simulator/Receivers/Receiver_PROG_A.cs
//////////////////////////////////////////////////////////
//
//  MK-52 RESURRECT
//  Copyright (c) 2020 Mike Yakimov.  All rights reserved.
//  See main file for the license
//
//////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace MK52Simulator
{
    //
    // Implements a generic empty receiver
    //
    public class Receiver_PROG_A: Receiver_PROG
    {
        public Receiver_PROG_A(MK52_Host parent)
            : base( parent)
        {
            Moniker = "PROG_A";
        }

        public override void activate(string prevReceiver)
        {
            base.activate(prevReceiver);
            LCD_Manager lm = _parent.getLCD();
            lm.updateStatusFMODE(" A ");
            lm.forcePaint();
        }

        public override byte tick(byte scancode)
        {
            RPN_Functions _rpnf = _parent.getFunctions();
            switch (scancode)
            {
                case 0: // keyboard inactive
                    return 0;

                // Column 0
                case 1:
                    return _rpnf.requestNextReceiver("PROG_F");
                case 2:
                    return _rpnf.requestNextReceiver("PROG_K");
                case 4:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_SET_DMOD_GRD);
                    break;

                // Column 1
                case 5:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_LBT);
                    return _rpnf.requestNextReceiver("TEXT", "PROG_N");
                case 6:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_LBZ);
                    return _rpnf.requestNextReceiver("TEXT", "PROG_N");
                case 7:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_LBY);
                    return _rpnf.requestNextReceiver("TEXT", "PROG_N");
                case 8:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_LBX);
                    return _rpnf.requestNextReceiver("TEXT", "PROG_N");

                // Column 2
                case 9:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_A_M2X);
                    return _rpnf.requestNextReceiver("ADDRESS", "PROG_N");
                case 10:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_A_X2M);
                    return _rpnf.requestNextReceiver("ADDRESS", "PROG_N");
                case 11:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_LBR);
                    break;
                case 12:
                    _rpnf.progMem.deleteLine();
                    break;

                // Column 3
                case 16:
                    _rpnf.progMem.commentLine();
                    return _rpnf.requestNextReceiver("TEXT", "PROG_N");

                // Column 5
                case 22:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_D2RAD);
                    break;
                case 23:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_MM2IN);
                    break;
                case 24:
                    return _rpnf.requestNextReceiver("FILE_N");

                // Column 6
                case 26:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_RAD2D);
                    break;
                case 27:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_IN2MM);
                    break;
                case 28:
                    return _rpnf.requestNextReceiver("DATA_N");

                // Column 7
                case 31:
                    _rpnf.appendProgramLine_P(RPN_Functions.FUNC_SEED);
                    break;
                case 32:
                case 33:
                    _rpnf.requestNextReceiver("PROG_N");
                    return 33;
                default: // all other buttons do nothing, keeping A-mode
                    return 0;
            }
            _rpnf.execute(RPN_Functions.FUNC_INCREMENT_PC);
            return _rpnf.requestNextReceiver("PROG_N");
        }
    }
}
